src/Tomcat/TomcatController.ts

```diff
@@ -293,12 +314,23 @@ export class TomcatController {
             watcher.close();
             if (serverInfo.needRestart) {
                 serverInfo.needRestart = false;
-                await this.startTomcat(serverInfo, appName);
+                await this.startTomcat(serverInfo);
             }
         } catch (err) {
             serverInfo.setStarted(false);
             if (watcher) { watcher.close(); }
             throw new Error(err.toString());
         }
     }
+    private async precheck(tomcatServer: TomcatServer): Promise<TomcatServer> {
+        if (!this._tomcatModel.getServerSet() || this._tomcatModel.getServerSet().length <= 0) {
```

Comment:
we can use isEmpty() in lodash.

I remember we have a lot of places that needs this kind of checking.

Leave the choice to you

Reply:
yes, will replace with lodash check in another PR to avoid this one being TL;DR